src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+                                       const std::string& gridName) {
+  assert(vector3Grids_.find(gridName) != vector3Grids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  vector3Grids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+std::shared_ptr<Mn::Trade::MeshData> VoxelGrid::getMeshData(
+    const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshDataDict_[gridName];
+}
+
+Mn::GL::Mesh& VoxelGrid::getMeshGL(const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshGLDict_[gridName];
+}
+
+Mn::Vector3 VoxelGrid::getGlobalCoords(const Mn::Vector3i& coords) {
+  Mn::Vector3 global_coords((coords[0]) * m_voxelSize[0],
+                            (coords[1]) * m_voxelSize[1],
+                            (coords[2]) * m_voxelSize[2]);
+  global_coords += m_offset;
+  return global_coords;
+}
+
+int VoxelGrid::generateBoolGridFromIntGrid(const std::string& intGridName,
+                                           std::string boolGridName,
+                                           int startRange,
+                                           int endRange) {
+  assert(intGrids_.find(intGridName) != intGrids_.end());
+  int num_filled = 0;
+  if (boolGridName == "")
+    boolGridName = intGridName;
+  addBoolGrid(boolGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                          m_voxelGridDimensions[2];
+       i++) {
+    if (intGrids_[intGridName].get()[i] >= startRange &&
+        intGrids_[intGridName].get()[i] <= endRange) {
+      boolGrids_[boolGridName].get()[i] = true;
+      num_filled++;
+    }
+  }
+  return num_filled;
+}
+
+int VoxelGrid::generateBoolGridFromVector3Grid(
+    const std::string& vector3GridName,
+    std::string boolGridName,
+    bool func(Mn::Vector3)) {
+  assert(vector3Grids_.find(vector3GridName) != vector3Grids_.end());
+  int num_filled = 0;
+  if (boolGridName == "")
+    boolGridName = vector3GridName;
+  addBoolGrid(boolGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
+                          m_voxelGridDimensions[2];
+       i++) {
+    if (func(vector3Grids_[vector3GridName].get()[i])) {
+      boolGrids_[boolGridName].get()[i] = true;
+      num_filled++;
+    }
+  }
+  return num_filled;
+}
+
+int VoxelGrid::generateBoolGridFromFloatGrid(const std::string& floatGridName,
+                                             std::string boolGridName,
+                                             float startRange,
+                                             float endRange) {
+  assert(floatGrids_.find(floatGridName) != floatGrids_.end());
+  int num_filled = 0;
+  if (boolGridName == "")
+    boolGridName = floatGridName;
+  addBoolGrid(boolGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
```

Comment:
This expression `m_voxelGridDimensions[0] * m_voxelGridDimensions[1] * m_voxelGridDimensions[2]` is used pretty often so let's have a helper `cellCount()` method or similar.